File: Is Subsequence/Is_Subsequence.py
# Runtime: 148 ms, faster than 67.76% of Python3 online submissions for Is Subsequence.
# Memory Usage: 14.3 MB, less than 100.00% of Python3 online submissions for Is Subsequence.

class Solution:
    def isSubsequence(self, s: 'str', t: 'str') -> 'bool':
        slength, tlength = len(s), len(t)

        if (slength == 0):
            return True
        
        sPtr, tPtr = 0, 0

        while tPtr < tlength:
            while tPtr < tlength and s[sPtr] != t[tPtr]:
                tPtr += 1
            if tPtr >= tlength:
                return False
            sPtr += 1
            if sPtr >= slength:
                return True
            tPtr += 1
        
        return False

File: Is Subsequence/Is_Subsequence.cpp
// 思路一，最为容易想到的线性时间复杂度的解决方案 
// Runtime: 48 ms, faster than 94.07% of C++ online submissions for Is Subsequence.
// Memory Usage: 17.1 MB, less than 100.00% of C++ online submissions for Is Subsequence.

class Solution 
{
public:
    bool isSubsequence(string s, string t) 
    {
        int slength = s.length(), tlength = t.length();
        if (slength == 0)
            return true;
        
        for  (int sPtr = 0, tPtr = 0; tPtr < tlength; tPtr++)
        {
            while (tPtr < tlength && s[sPtr] != t[tPtr])
                tPtr++;
            if (tPtr >= tlength)
                return false;
            if (++sPtr >= slength)
                return true;    
        }
        return false;
    }
};